sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/planner/BeamRuleSets.java

```diff
@@ -103,6 +106,10 @@
 
           // join rules
           JoinPushExpressionsRule.INSTANCE,
+          JoinCommuteRule.INSTANCE,
```

Comment:
Is there a good way to test the effect of adding such rules. We have seen examples [1] and [2] that adds rules that causes issues.

[1]: https://github.com/apache/beam/pull/8035
[2]: https://github.com/apache/beam/pull/6139

Reply:
cc @kanterov as well who also had relevant experience.

Reply:
I have tested the effect of adding these rules in terms of performance gain for some of the queries in https://docs.google.com/document/d/1DM_bcfFbIoc_vEoqQxhC7AvHBUDVCAwToC8TYGukkII/edit#

However, in terms of testing against bugs, all I can say is it is passing our current tests; therefore, if it can create illegal plans for some query, that means we are not covering it in our tests and it should be added. I am not sure how we can test this against all possible scenarios and queries. I think one way to improve our test cases is checking other systems' tests (Such as Calcite's test cases) and add them to ours. I will appreciate if you have any suggestion for that

Reply:
But it's less common to have a change without any test?

I am thinking a test that you can enable rules and then check output plan.

Reply:
In general we should have tests. However in this case the inputs/outputs are not changing depending on the inputs. We should make sure that we have a couple of tests that we know trigger these rules but we cannot enforce it, i.e. I don't know how we can ensure that we even hit the rule by looking at the results. If we don't know what a rule breaks, can we come up with a meaningful test for it?

I also don't think we should check the plans. It will be too brittle. Rules can be enabled and disabled independently and probably interact in hard-to-predict ways. If VolcanoPlanner (or in general Calcite stuff) is tested, then we should not re-test it and accept that it applies the rules correctly. We add some rules that are applied the same way as all other rules and inherit from them. In this case the rules behave the same way as built-in Calcite rules, and reject the plans that we already don't support in the join rel.

It feels like the only issue can be is the applicability of the reordering operation itself. Looking at what join operations we support and how it works (e.g. window is inherited for non-merging windows, or the input has to be explicitly re-windowed for merging windows, and that we only allow the default trigger that has itself as a continuation trigger), it looks safe and from the top of the head I cannot come up with a test case that we accept but that would break with the rule / without the rule.

Probably a unit test of the join rejection would make sense but I don't think it's  really valuable as we literally call the same code that we already do in the existing join rel and that probably already have tests (this should be confirmed) that trigger that logic.

Reply:
Give me some time that let me check the details of PR and how a useful test could be.

Generally if we are making changes that we cannot test, we should postponing such changes until we figure out how to test it.

I think planner rules can be tested because I see Calcite and Flink have tests around optimization rules.

Reply:
Because this PR is implementing reordering joins, a useful test would be a test, in which a three-way join is reordered. As join reordering is already measured in https://docs.google.com/document/d/1DM_bcfFbIoc_vEoqQxhC7AvHBUDVCAwToC8TYGukkII/edit#, wouldn't it be straightforward to have a similar test? Without such a test, how do we even know if join reordering is working?

In terms of checking output plans, Flink has been doing many tests on Calcite optimization rules (see [here](https://github.com/apache/flink/blob/master/flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/plan/rules/logical/RewriteMultiJoinConditionRuleTest.scala) and [here](https://github.com/apache/flink/tree/master/flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/plan)). Flink's practice has shown that verifying output plan is deterministic and stable.

The basic idea is if you want to test an optimization, only enable relevant rules in test case(so rules are hit will be known) I can see by Flink's way, you can test rules even if rules can be disabled and enabled independently:

In [BeamRuleSets.java](https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/planner/BeamRuleSets.java)
```
static BeamJoinReorderingRelSet = 
{JoinCommuteRule.INSTANCE, 
JoinAssociateRule.INSTANCE}
```

In JoinReorderingTest.java
```
FrameworkConfig testConfig = createTestConfig(BeamJoinReorderingRelSet)
Planner testPlanner = Frameworks.getPlanner(testConfig);

// setup input tables

expected_plan = PlanLoader.load(testcase.class)
verifyPlan(testPlanner.getPlan(sql), expected_plan)
```

By doing so, if a relevant rule is disable, it will break existing join reordering tests, which has good value for us. It also justifies this PR is doing join reordering.

Because we are starting the effort to have more optimization rules in BeamSQL, Flink's practice on testing is a great example that we can learn and apply to BeamSQL to maintain our codebase's health.

Reply:
I just added a test to check if the join of three tables are reordered in the expected way.

Reply:
There are few separate kinds of tests relevant here:

* tests of Beam/Beam SQL:
  * we test that the system works correctly;
  * we verify that expected inputs produce correct results;
     * as long as everything works correctly and produces correct data, everything is good;
  * if we add a new optimization rule that doesn't affect the behavior+correctness of the system, no tests in this category should fail;

* tests of specific rules. They test that a concrete rule works as expected. To test it we need to make sure that:
  * without the rule the behavior doesn't happen (no reorders);
  * with the rule the behavior happens (reorder does happen);
  * this specific rule is applied and caused the change, and no other rule could have caused this effect:
    * this potentially can happen if there are multiple interdependent rules that trigger in presence of each other. Whether this is a correct and intended situation or not we should be isolating this case;
  * if someone adds another rule to the bigger "production" set, this suite should not break, it only tests the specific rule;
  * if we decide to not use the rule and exclude it from the "production" rule set, this test should not break;

* "integration" test, or system configuration test. We ensure that the system is configured the certain way:
   * we can look at what rules are enabled and fail if the enabled rule set is different from what we expect;
   * we can test that the larger "production" ruleset has some behavior, e.g. that all rules combined have the effect of ordering enabled, so that if someone disables it with another rule, it breaks;

I am not sure if we currently have a straightforward way to setup the tests that exercise only one rule. E.g. to test a concrete rule we probably need to setup the parser and make it work without all rules except the ones we need, probably even without Beam physical rules (which we also don't test directly). If we are able to get something like this from Calcite/Flink, we should do it.